subsys/mgmt/mcumgr/smp_client/src/client.c

```diff
@@ -0,0 +1,326 @@
+{
+	struct smp_hdr smp_header;
+	enum mcumgr_op_t response;
+	struct smp_client_cmd_req *cmd_req;
+
+	if (sys_slist_is_empty(&smp_client_data.cmd_list)) {
+		return NULL;
+	}
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&smp_client_data.cmd_list, cmd_req, node) {
+		smp_read_hdr(cmd_req->nb, &smp_header);
+		if (smp_header.nh_op == MGMT_OP_READ) {
+			response = MGMT_OP_READ_RSP;
+		} else {
+			response = MGMT_OP_WRITE_RSP;
+		}
+
+		if (smp_header.nh_seq != res_hdr->nh_seq) {
+			continue;
+		} else if (res_hdr->nh_op != response) {
+			continue;
+		}
+
+		return cmd_req;
+	}
+	return NULL;
+}
+
+int smp_client_object_init(struct smp_client_object *smp_client, int smp_type)
+{
+	smp_client->smpt = smp_client_transport_get(smp_type);
+	if (!smp_client->smpt) {
+		return MGMT_ERR_EINVAL;
+	}
+
+	/* Init TX FIFO */
+	k_work_init(&smp_client->work, smp_client_handle_reqs);
+	k_fifo_init(&smp_client->tx_fifo);
+
+	return MGMT_ERR_EOK;
+}
+
+int smp_client_single_response(struct net_buf *nb, const struct smp_hdr *res_hdr)
+{
+	struct smp_client_cmd_req *cmd_req;
+	smp_client_res_fn cb;
+	void *user_data;
+
+	/* Discover request for incoming response */
+	cmd_req = smp_client_response_discover(res_hdr);
+	LOG_DBG("Response Header len %d, flags %d OP: %d group %d id %d seq %d", res_hdr->nh_len,
+		res_hdr->nh_flags, res_hdr->nh_op, res_hdr->nh_group, res_hdr->nh_id,
+		res_hdr->nh_seq);
+
+	if (cmd_req) {
+		cb = cmd_req->cb;
+		user_data = cmd_req->user_data;
+		smp_client_cmd_req_free(cmd_req);
+		if (cb) {
+			cb(nb, user_data);
+			return MGMT_ERR_EOK;
+		}
+	}
+
+	return MGMT_ERR_ENOENT;
+}
+
+struct net_buf *smp_client_buf_allocation(struct smp_client_object *smp_client, uint16_t group,
+					  uint8_t command_id, uint8_t op,
+					  enum smp_mcumgr_version_t version)
+{
+	struct net_buf *nb;
+	struct smp_hdr smp_header;
+
+	nb = smp_packet_alloc();
+
+	if (nb) {
+		/* Write SMP header with payload length 0 */
+		smp_header_init(&smp_header, group, command_id, op, 0, smp_client->smp_seq++,
+				version);
+		memcpy(nb->data, &smp_header, sizeof(smp_header));
+		nb->len = sizeof(smp_header);
+	}
+	return nb;
+}
+
+void smp_client_buf_free(struct net_buf *nb)
+{
+	smp_packet_free(nb);
+}
+
+static void smp_read_hdr(const struct net_buf *nb, struct smp_hdr *dst_hdr)
+{
+	memcpy(dst_hdr, nb->data, sizeof(*dst_hdr));
+	dst_hdr->nh_len = sys_be16_to_cpu(dst_hdr->nh_len);
+	dst_hdr->nh_group = sys_be16_to_cpu(dst_hdr->nh_group);
+}
+
+int smp_client_send_cmd(struct smp_client_object *smp_client, struct net_buf *nb,
+			smp_client_res_fn cb, void *user_data, int timeout_in_sec)
+{
+	struct smp_hdr smp_header;
+	struct smp_client_cmd_req *cmd_req;
+
+	if (timeout_in_sec > 30) {
+		LOG_ERR("Command timeout can't be over 30 seconds");
+		return MGMT_ERR_EINVAL;
+	}
+
+	if (timeout_in_sec == 0) {
+		timeout_in_sec = CONFIG_SMP_CMD_DEFAULT_LIFE_TIME;
+	}
+
+	smp_read_hdr(nb, &smp_header);
+	if (nb->len > sizeof(smp_header)) {
```

Comment:
Would flip this to error out if there is not at least `sizeof(smp_header)` to send, as that would not be valid, then the part below can be part of the main flow as it will only send packets of a minimum valid length